### PR TITLE
Initial commit of rough documentation (resolves #511)

### LINKED ISSUE #511: Document job.py

### PR DESCRIPTION
**Pull request by benedictpaten**
Includes:
an overview of toil features
public functions of job.py
a high level tutorial of toil features
stubs for the job store and batch system classes

Resolves #511

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Added this branch to RTD:

http://toil.readthedocs.org/en/issues-511-revised-user-documentation/

**Comment by hannes-ucsc** [comment]
Looks great. A lot of meat. I think the word `job` should be lowercase since its not a proper noun. When referring to the `Job` class, double backqoutes should be used. For example:

```
The ``Job`` class encapsulates information about *jobs*. A *job* is the basic unit of work in Toil.
```

This renders as

> The `Job` class encapsulates information about *jobs*. A *job* is the basic unit of work in Toil.

I can do the substitutions, just let me know how passionate you are about Job vs job.

Also, I don't think the backslashes at EOL in the docstrings are necessary. 

Request to go multi-author with this one.

**Comment by hannes-ucsc** [comment]
Looks like the job vs Job thing is only on the first page.

**Comment by hannes-ucsc** [comment]
Yes. See also the existing tickets on documentation. Lets get this merged as quickly as possible. I want to do reformatting and package reorg stuff soon.

Does your IDE have a function for hard-wrapping paragraphs to a fixed line length? Both of the tools I use have that (PyCharm and TextMate). I suggest that we wrap the paragraphs docstring to 100 in Toil source and to 80 in plain .rst files.

**Comment by benedictpaten** [comment]
I'm assigning this to Jeltje to do a review. There are rough edges. I agree, which should get this merged as soon as possible (and Jeltje has time). If Jeltje does not have time now I'd propose assigning to someone else.

File: docs/toilAPI.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Job.Service
+-----------
+The Service class allows databases and servers to be spawned within a Toil workflow.
+
+.. autoclass:: toil.job::Job.Service
+   :members:
+   
+FunctionWrappingJob
+-------------------
+The subclass of Job for wrapping user functions.
```
Comment (Jeltje) [inline docs/toilAPI.rst]: FIXME: Description needs more detail for clarity

File: docs/toilAPI.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+.. autoclass:: toil.job::FunctionWrappingJob
+   :members:
+   
+JobFunctionWrappingJob
+----------------------
+The subclass of FunctionWrappingJob for wrapping user job functions.
```
Comment (Jeltje) [inline docs/toilAPI.rst]: FIXME: Needs more detail on difference with FunctionWrappingJob (the naming is confusing)

File: docs/tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+specified by methods of the job class, e.g. :func:`toil.job.Job.addChild` 
+and :func:`toil.job.Job.addFollowOn`. 
+
+Considering a set of jobs the nodes in a job graph and the child and follow-on 
+relationships the directed edges of the graph, we say that a job B that is on a directed 
+path of child/followOn edges from a job A in the job graph is a *successor* of A, 
+similarly A is a *predecessor* of B.
+
+A parent job's child jobs are run directly after the parent job has completed, and in parallel. 
+The follow-on jobs of a job are run after it's child jobs and their successors
```
Comment (Jeltje) [inline docs/tutorial.rst]: its (not it's)

File: docs/tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        # Read the second file again using a stream.
+        with job.fileStore.readGlobalFileStream(fileID2) as fH:
+            print fH.read() #This prints "Out brief candle"
+        
+        # Delete the first file from the global file store.
+        job.fileStore.deleteGlobalFile(fileID)
+        
+        # It is unnecessary to delete the file keyed by fileID2 
+        # because we used the cleanup flag, which removes the file after this 
+        # job and all it's successors have run (if the file still exists)
```
Comment (Jeltje) [inline docs/tutorial.rst]: its (not it's)

File: docs/tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+In this example the DemoService starts a database in the start method,
+returning an object from the start method indicating how a client job would access the database. 
+The service's stop method cleans up the database.
+
+A DemoService instance is added as a service of the root job *j*, with resource requirements
+specified. The return value from :func:`toil.job.Job.addService` is a promise to the return
+value of the service's start method. When the promised is fulfilled it will represent how
+to connect to the database. The promise is passed to a child job of j, which
+uses it to make a database connection. The services of a job are started before any of 
+it's successors have been run and stopped after all the successors of the job have completed
```
Comment (Jeltje) [inline docs/tutorial.rst]: its (not it's)

File: docs/tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+runs after A and all its successors is to create a parent of A, call it Ap, \
+and then make B a follow-on of Ap. e.g.::
+    from toil.job import Job
+    
+    # A is a job with children and follow-ons, for example:
+    A = Job()
+    A.addChild(Job())
+    A.addFollowOn(Job())
+    
+    # B is a job which needs to run after A and it's successors
```
Comment (Jeltje) [inline docs/tutorial.rst]: its (not it's)

File: docs/tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    
+    # A 
+    A = Job()
+    A.addChild(Job())
+    A.addFollowOn(Job())
+    
+    #Encapsulate A
+    A = A.encapsulate()
+    
+    # B is a job which needs to run after A and it's successors
```
Comment (Jeltje) [inline docs/tutorial.rst]: its (not it's)

File: src/toil/job.py [hunk 8]
```diff
@@ -1231,33 +1361,71 @@ def _jobName(self):
...
         self.fileStore = fileStore
         rValue = userFunction(*((self,) + tuple(self._args)), **self._kwargs)
         return rValue
 
+class EncapsulatedJob(Job):
+    """
+    A convenience Job class used to make a job subgraph appear to be a single job. 
+    
+    Let A be the root job of a job subgraph and B be another job we'd like to run after A
+    and all it's successors have completed, for this use encapsulate::
```
Comment (Jeltje) [inline src/toil/job.py]: its (not it's)

File: src/toil/job.py [hunk 9]
```diff
@@ -1231,33 +1361,71 @@ def _jobName(self):
...
+        A, B = A(), B() #Job A and subgraph, Job B
+        A' = A.encapsulate()
+        A'.addChild(B) #B will run after A and all it's successors have
```
Comment (Jeltje) [inline src/toil/job.py]: its (not it's)

File: src/toil/job.py [hunk 10]
```diff
@@ -1231,33 +1361,71 @@ def _jobName(self):
...
+class EncapsulatedJob(Job):
+    """
+    A convenience Job class used to make a job subgraph appear to be a single job. 
+    
+    Let A be the root job of a job subgraph and B be another job we'd like to run after A
+    and all it's successors have completed, for this use encapsulate::
+        A, B = A(), B() #Job A and subgraph, Job B
+        A' = A.encapsulate()
+        A'.addChild(B) #B will run after A and all it's successors have 
+        # completed, A and it's subgraph of successors in effect appear
```
Comment (Jeltje) [inline src/toil/job.py]: its (not it's)

File: docs/index.rst [hunk 11]
```diff
@@ -3,63 +3,67 @@
...
-follow on is the order of execution. Parents are executed first, its children, finally
-followed by the follow ons. The children and follow on jobs are run in parallel.
+The workflow consists of a single job, which calls the helloWorld function. The resource
+requirements for that job are (optionally) specified by keyword arguments (memory, cores, disk).
 
+The :class:`toil.job.Job.Runner` class handles the invocation of Toil workflows. 
+It is fed an options object that configures the running the workflow. 
+This can be populated by an argument parser object using 
+:func:`toil.job.Job.Runner.getDefaultArgumentParser`, allowing all these options to specified 
+via the command line to the script.
```
Comment (Jeltje) [inline docs/index.rst]: Link to the tutorial?

File: docs/jobStore.rst [hunk 12]
```diff
@@ -0,0 +1,21 @@
+The job store interface
+=======================
+
+Implementing the job store interface tutorial
+*********************************************
+
+The job store interface is used by Toil to abstract over different file stores,
```
Comment (Jeltje) [inline docs/jobStore.rst]: How about something like 
`The job store interface is an abstraction layer that that hides the specific details of file sytem storage`

I think 'abstract over' is too.. abstract.

File: src/toil/job.py (hunk 10 above)
Comment (benedictpaten) [inline src/toil/job.py]: Apparently I make that typo a lot :)

**Comment by hannes-ucsc** [comment]
@Jeltje Feel free to fix obvious typos directly in the code and push them to this branch. I would use the line notes only for things that you are not sure how to fix yourself.

**Comment by benedictpaten** [comment]
Yes, I'm happy to fix these this time as I make my second pass, but as a
policy I think it's easier in the future for people to just fix for
themselves. Really appreciate the other comments, btw.

**Comment by Jeltje** [comment]
Hey all, 
I'm going to fix and push the spelling errors so Benedict can focus on my other comments.

**Comment by Jeltje** [comment]
Done with typos. Comment from email, moving here:

There are objects in the job library that are not mentioned in toilAPI.rst (JobGraphDeadlockException and PromisedJobReturnValue) - I think those fall under Execptions and Promise, which are listed (and are presumably stubs).

File: docs/toilAPI.rst (hunk 1 above)
Comment (benedictpaten) [inline docs/toilAPI.rst]: I'm a little unclear what you want to see here. There is documentation with the toil/job.py module for each of these that is embedded here. 

I've tended to not duplicate documentation, and tried to cover the use of the functions in the tutorial, so I'd prefer not to duplicate too much, but (obviously) I do want to make sure it's clear.

File: docs/toilAPI.rst (hunk 2 above)
Comment (benedictpaten) [inline docs/toilAPI.rst]: Agree we could do with a better name, but I don't want to fix that in this PR. I think the inline documentation in the class itself does an okay job of documenting what a job function is, it is also fairly covered in the tutorial?

File: docs/jobStore.rst (hunk 12 above)
Comment (benedictpaten) [inline docs/jobStore.rst]: Yes, much better, I fixed this.

File: docs/index.rst (hunk 11 above)
Comment (benedictpaten) [inline docs/index.rst]: Yes, adding this.

**Comment by benedictpaten** [comment]
Okay, I'm going to merge once the tests pass.

**Pull request merged**